test/lit/passes/simplify-globals-read_only_to_write.wast

```diff
@@ -386,4 +352,187 @@
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $side-effects-in-condition-2
+    (if
+      (if (result i32)
+        (call $foo) ;; these side effects are not a problem, as the global's
+                    ;; value cannot reach them.
+        (i32.const 1)
+        (global.get $global) ;; the global's value flows out through the if,
+                             ;; safely
+      )
+      (global.set $global (i32.const 1))
+    )
+  )
+
+  ;; CHECK:      (func $foo (result i32)
+  ;; CHECK-NEXT:  (unreachable)
+  ;; CHECK-NEXT: )
+  (func $foo (result i32)
+    (unreachable)
+  )
+)
+
+;; As above, but now the global's value flows into a side effect.
+(module
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (global $global (mut i32) (i32.const 0))
+  (global $global (mut i32) (i32.const 0))
+  ;; CHECK:      (global $other i32 (i32.const 0))
+  (global $other (mut i32) (i32.const 0))
+  ;; CHECK:      (func $side-effects-in-condition-3
+  ;; CHECK-NEXT:  (local $temp i32)
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (local.tee $temp
+  ;; CHECK-NEXT:    (global.get $global)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (global.set $global
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $side-effects-in-condition-3
+    (local $temp i32)
+    (if
+      (local.tee $temp
+        (global.get $global) ;; the global's value flows into a place that has
+      )                      ;; side effects, so it may be noticed.
+      (global.set $global (i32.const 1))
+    )
+  )
+)
+
+;; As above, but now the global's value flows through multiple layers of
+;; things that have no side effects and are safe.
+(module
+  (memory 1 1)
+
+  ;; CHECK:      (type $none_=>_none (func))
+
+  ;; CHECK:      (global $once i32 (i32.const 0))
+  (global $once (mut i32) (i32.const 0))
+
+  ;; CHECK:      (memory $0 1 1)
+
+  ;; CHECK:      (func $side-effects-in-condition-4
+  ;; CHECK-NEXT:  (local $x i32)
+  ;; CHECK-NEXT:  (local $y i32)
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (i32.eqz
+  ;; CHECK-NEXT:    (select
+  ;; CHECK-NEXT:     (local.tee $x
+  ;; CHECK-NEXT:      (i32.const 1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (i32.load
+  ;; CHECK-NEXT:      (i32.const 2)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (i32.add
+  ;; CHECK-NEXT:      (i32.const 0)
+  ;; CHECK-NEXT:      (i32.const 1337)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:   (drop
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $side-effects-in-condition-4
+    (local $x i32)
+    (local $y i32)
+    (if
+      (i32.eqz
+        (select
+          (local.tee $x
+            (i32.const 1)
+          )
+          (i32.load
+            (i32.const 2)
+          )
+          (i32.add
+            (global.get $once)
```

Comment:
It might be fun to sprinkle multiple reads of the global in here as well.

Reply:
That should work in principle, yeah. But it won't work atm, as we count the number of globals. We basically want to see that the number of reads of the global is equal to the number of places where it is "read but only to be written". Then we subtract the latter, which are the "safe" reads that we understand cause no other side effects than writing the global. If any other reads remain then we must assume the worst.

I do actually have a followup for this that I'll post once this lands, which now that I think of it may also handle that case, but I'll need to check.